robottelo/performance/perf_candlepin.py

```diff
@@ -0,0 +1,116 @@
+and have utilities of single register by activation-key, single
+register and attach, single subscription deletion.
+
+"""
+import logging
+import time
+import requests
+
+from robottelo.common import conf, ssh
+from robottelo.common.helpers import get_server_credentials, get_server_url
+from urlparse import urljoin
+
+
+class CandlepinPerfTest(object):
+    """Measures performance of RH Satellite 6
+
+    Candlepin Subscription functionality
+
+    """
+
+    logger = logging.getLogger('robottelo')
+
+    # parameters for Candlepin Authentication
+    username = conf.properties['foreman.admin.username']
+    password = conf.properties['foreman.admin.password']
+    serverhost = conf.properties['main.server.hostname']
+
+    requests.packages.urllib3.disable_warnings()
+
+    @classmethod
+    def single_register_activation_key(cls, ak_name, default_org, vm):
+        """Subscribe VM to Satellite by Register + ActivationKey"""
+
+        # note: must create ssh keys for vm if running on local
+        result = ssh.command('subscription-manager clean', hostname=vm)
+        result = ssh.command(
+            'time -p subscription-manager register '
+            '--activationkey={0} --org={1}'
+            .format(ak_name, default_org), hostname=vm)
+
+        if result.return_code != 0:
+            cls.logger.error('Fail to register by activation-key!')
+        else:
+            real_time = [real for real in result.stderr.split('\n')
+                         if real.startswith('real')]
+            real_time = real_time[0].split(' ')[1]
+            return float(real_time)
+
+    @classmethod
+    def single_register_attach(cls, sub_id, default_org, environment, vm):
+        """Subscribe VM to Satellite by Register + Attach"""
+
+        ssh.command('subscription-manager clean', hostname=vm)
+        time_reg = cls.sub_mgr_register_authentication(
+            default_org, environment, vm)
+
+        time_att = cls.sub_mgr_attach(sub_id, vm)
+
+        if time_reg is None or time_att is None:
+            total_time = 0
+        else:
+            total_time = time_reg + time_att
+        return time_reg, time_att, total_time
+
+    @classmethod
+    def sub_mgr_register_authentication(cls, default_org, environment, vm):
+        """subscription-manager register -u -p --org --environment"""
+        result = ssh.command(
+            'time -p subscription-manager register --username={0} '
+            '--password={1} --org={2} --environment={3}'
+            .format(cls.username, cls.password, default_org, environment),
+            hostname=vm)
+
+        if result.return_code == 0:
+            real_time = [real for real in result.stderr.split('\n')
+                         if real.startswith('real')]
+            real_time = real_time[0].split(' ')[1]
+            return float(real_time)
+        else:
+            cls.logger.error('Fail to register for subscription manager!')
```

Comment:
This is equivalent:

```python
        if result.return_code != 0:
            cls.logger.error('Fail to register for subscription manager!')
            return
        real_time = [
            real
            for real in result.stderr.split('\n')
            if real.startswith('real')
        ]
        return float(real_time[0].split(' ')[1])
```

One technique that almost always produces nicer Python code is to adopt a "check
for failure" mindset, not a "check for success" mindset. Why? Consider the
difference between this:

    if foo == 'desired result':
        do things
        if bar == 'desired result':
            do many things
        else:
            briefly handle the error
    else:
        briefly handle the error

And this:

    if foo != 'desired result':
        briefly handle the error
    do things
    if bar != 'desired result':
        briefly handle the error
    do many things

The latter is shorter and less indented.